pallets/xcm-core-buyer/src/lib.rs

```diff
@@ -640,44 +733,31 @@ pub mod pallet {
         type Call = Call<T>;
 
         fn validate_unsigned(_source: TransactionSource, call: &Self::Call) -> TransactionValidity {
-            if let Call::buy_core { para_id, proof } = call {
-                /*
-                if <Pallet<T>>::is_online(heartbeat.authority_index) {
-                    // we already received a heartbeat for this authority
-                    return InvalidTransaction::Stale.into()
+            if let Call::buy_core {
+                para_id,
+                collator_account_id,
+                proof,
+            } = call
+            {
+                let block_number = <frame_system::Pallet<T>>::block_number();
+
+                let current_nonce = CollatorSignatureNonce::<T>::get();
```

Comment:
Why do we need a nonce? Can't we use the block number as the nonce for example?

Reply:
Hmm. Yeah. We can use block number as nonce. But problem is that it is not guaranteed at which block this tx will be included and any tx signed with earlier block number can become invalid. 

So, the idea is to have nonce just like a substrate account but shared between all calls.

Reply:
I don't like that it's an extra storage map. But block number won't work, because the tx may not be included in the next block, you're right. Container chain block number (from author-noting pallet) also won't work, because in case of failure we won't be able to retry... There must be something that we can use though.

Reply:
It is not extra storage map. It's just one storage value at the moment. 

I think it is nice trade off between amount of space and the uncertainity.

Reply:
Oh, really? Sorry, I haven't look into the code yet, but what will happen if two collators try to buy a core in the same block? Will it work or will it fail?

Reply:
One of them would succeed and other one would fail. That is sort of drawback. This is the reason why I wanted it originally as storage map.

On second thought, let me modify it as storage map first and then we can arrive at equivalent solution.

Reply:
Yeah I agree, let's work with a storage map now, write some tests to make sure everything works, and then we can think about how to remove this storage map.